### PR TITLE
introduce a method for fetching relay chain header to `RelayChainInterface`

### PR DESCRIPTION
**Pull request by seunlanlege**
This is particularly useful for fetching arbitrary relay chain headers within inherents.

### CODE REVIEW

**Review by BradleyOlson64 — APPROVED** [review]

**Comment by BradleyOlson64** [comment]
Looks like @rphmeier introduced a method with the same intent and a slightly different signature [here](https://github.dev/paritytech/cumulus/blob/e00621fef26bbaf9ebdd6646936101a21e696ac2/client/relay-chain-interface/src/lib.rs#L117).

Perhaps your version could be used instead.

**Comment by seunlanlege** [comment]
Yeah my use case requires block numbers

File: client/relay-chain-interface/src/lib.rs [hunk 1]
```diff
@@ -110,6 +111,9 @@ pub trait RelayChainInterface: Send + Sync {
 	/// Get the hash of the current best block.
 	async fn best_block_hash(&self) -> RelayChainResult<PHash>;
 
+	/// Fetch the block header of a given height
+	async fn header(&self, block_id: BlockId) -> RelayChainResult<PHeader>;
```
[the remark this answers is not in the document]
Comment (seunlanlege) [reply client/relay-chain-interface/src/lib.rs]: fixed

File: client/relay-chain-interface/src/lib.rs [hunk 2]
```diff
@@ -110,6 +111,9 @@ pub trait RelayChainInterface: Send + Sync {
 	/// Get the hash of the current best block.
 	async fn best_block_hash(&self) -> RelayChainResult<PHash>;
 
+	/// Fetch the block header of a given height
+	async fn header(&self, block_id: BlockId) -> RelayChainResult<Option<PHeader>>;
```
Comment (bkchr) [inline client/relay-chain-interface/src/lib.rs:115]: Please use `PHash`. No new `BlockId` in public interfaces anymore.

Comment (seunlanlege) [reply client/relay-chain-interface/src/lib.rs:115]: But i specifically need to query using `BlockNumber`, any reason for this? 

> No new BlockId in public interfaces anymore

Comment (bkchr) [reply client/relay-chain-interface/src/lib.rs:115]: Block numbers should only be used in places where there is no other way around them. Block numbers are not ambiguous. Some interfaces (like RPC/CLI) still require block numbers, but most "internal code" can be rewritten in a way to use hashes. E.g. get the header of block X and then go back using `parent_hash()`.

Comment (seunlanlege) [reply client/relay-chain-interface/src/lib.rs:115]: > But i specifically need to query using BlockNumber, any reason for this?

The use case here is on-chain code which wishes to read the relay chain, they signal a request given a specific relay chain height (parachains don't have access to the relay chain blockhashes). So hence the `BlockId` here. 

Doesn't seem like there's any alternatives here

Comment (bkchr) [reply client/relay-chain-interface/src/lib.rs:115]: And how does the parachain validates the relay chain headers?

Comment (bkchr) [reply client/relay-chain-interface/src/lib.rs:115]: > The use case here is on-chain code which wishes to read the relay chain, they signal a request given a specific relay chain height (parachains don't have access to the relay chain blockhashes).

You say it signals which relay chain header it wants to read. So you are just wanting to get the header to get the storage root?

Comment (seunlanlege) [reply client/relay-chain-interface/src/lib.rs:115]: yes

Comment (bkchr) [reply client/relay-chain-interface/src/lib.rs:115]: Okay, when we got: https://github.com/paritytech/cumulus/issues/303

This can be changed.

**Review by bkchr — APPROVED** [review]

**Comment by rphmeier** [comment]
This shouldn't have been merged as-is - the interface was changed to `Option<Header>` but the implementations still wrapped it in opaque errors for `BlockNumber`

**Comment by bkchr** [comment]
By "opaque error" you mean that it is a string error?

**Comment by seunlanlege** [comment]
Will do a follow up

**Comment by seunlanlege** [comment]
follow up: https://github.com/paritytech/cumulus/pull/2830

**Pull request merged**